File: practice_in_java/multithreading/Primes.java
package multithreading;

// import interface_package.myinterface;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.*;
import java.util.*;

public class Primes
{
    // initial value of 1 to account for the 2 being only even prime number
    // will keep track of all prime numbers between 1 and 100 million
    private static AtomicInteger cnt = new AtomicInteger(1);

    // initial value of 2 to account for the 2 being only even prime number
    // will keep track of the sum of all prime numbers between 1 and 100 million
    private static AtomicLong sum = new AtomicLong(2L);

    
    private static List<Integer> primeNumbers = Collections.synchronizedList(new ArrayList<Integer>());

    private static void CheckThreadCompletion(Thread[] threads)
    {
        boolean allThreadsAreDone = false;
        while (!allThreadsAreDone)
        {
            for (Thread t : threads)
            {
                if (t.isAlive())
                {
                    allThreadsAreDone = false;
                    break;
                }
                allThreadsAreDone = true;
            }
            // break will take us here
        }
    }

    // Run each thread one at a time
    private static void RunDemThreads(Thread[] sieveThreads, UseSieves[] cPrimes)
    {
        for (int i = 0; i < sieveThreads.length; i++)
        {
            sieveThreads[i] = new Thread(cPrimes[i]);
            sieveThreads[i].start();
            try
            {
                sieveThreads[i].join();
            } catch (Exception e)
            {
                System.out.println("exception caught");
            }
        }
    }

    // This method sets the starting points for each thread
    private static void CreateThreadWorkers(UseSieves[] cPrimes, int upperBound, boolean[] primes, int[] primeNums)
    {
        for (int i = 0; i < cPrimes.length; i++)
        {
            // create 8 calculateprimes objects each with unique start and ends
            cPrimes[i] = new UseSieves(primeNums[i], upperBound, primes);
        }
    }

    private static void TheFunctionThatDoesEverything()
    {
        int numThreads = 8;
        Thread[] sieveThreads = new Thread[numThreads];
        UseSieves[] cPrimes = new UseSieves[numThreads];
        int upperBound = 100000000;
        boolean[] primes = new boolean[upperBound + 1];
        int[] primeNums = new int[]
        { 3, 5, 7, 11, 13, 17, 31, 41 };
        Arrays.fill(primes, true);
        primes[0] = primes[1] = false;

        // create the threads
        CreateThreadWorkers(cPrimes, upperBound, primes, primeNums);

        // run the threads
        RunDemThreads(sieveThreads, cPrimes);

        // let all the threads process then get all the primes afterwards
        CheckThreadCompletion(sieveThreads);

        // write code here to loop thru 10^8 to get all primes

        ExecutorService ex = Executors.newFixedThreadPool(numThreads);

        var lstofStarts = new int[]
        { 0, 12500000, 25000000, 37500000, 50000000, 62500000, 75000000, 87500000, };

        for (var num : lstofStarts)
        {
            ex.execute(new Runnable()
            {
                private void GetFinalizedContents(int start, boolean[] primes)
                {
                    // #8 threads will count all the primes in its segment
                    // for 1e8, the first thread will count from 0 to 12.5 million, exclusive;
                    // #the second thread will count from 12.5 million to 25 million, exclusive;
                    // #and so on and so forth
                    int offset = 12500000;
                    for (int i = start + 1; i < start + offset; i += 2)
                    {
                        if (primes[i])
                        {
                            sum.addAndGet(i);
                            cnt.incrementAndGet();
                            primeNumbers.add(i);
                        }
                    }
                }

                @Override
                public void run()
                {
                    GetFinalizedContents(num, primes);
                }
            });
        }

        // Wait for all threads to finish their work and join up
        ex.shutdown();
        while (!ex.isTerminated())
        {
            Thread.currentThread().setPriority(Thread.MIN_PRIORITY);
            Thread.yield();
        }

        // #region old way of getting all primes
        // ArrayList<Integer> a = new ArrayList<>();
        // a.add(2);
        // Long sum = 2L;
        // for (int i = 3; i < upperBound; i += 2)
        // {
        // if (primes[i])
        // {
        // a.add(i);
        // sum += i;
        // }
        // }

        // int len = a.size();
        // System.out.println("total sum: " + sum);
        // System.out.println("number of primes: " + len);
        // System.out.println("ten largest primes: " + a.subList(len - 10, len));
        // #endregion
    }

    public static void main(String[] args)
    {
        long start = System.currentTimeMillis();
        TheFunctionThatDoesEverything();

        System.out.println(cnt);
        System.out.println(sum);

        Collections.sort(primeNumbers);
        primeNumbers.subList(primeNumbers.size() - 10, primeNumbers.size()).forEach(i -> System.out.print(i + " "));
        System.out.println();
        System.out.println("Total runtime (in milliseconds): " + (System.currentTimeMillis() - start));
    }
}

class UseSieves implements Runnable
{
    private int startInd, endInd;
    private boolean[] primes;

    public UseSieves(int startInd, int endInd, boolean[] primes)
    {
        this.startInd = startInd;
        this.endInd = endInd;
        this.primes = primes;
    }

    public synchronized void SieveEmForGood()
    {
        // caching should increase efficiency rather than repeatedly
        // making the function call
        int res = (int) Math.sqrt(endInd);
        for (int i = startInd; i < res; i += 2)
        {
            if (primes[i])
            {
                for (int j = i * i; j < endInd; j += i)
                {
                    primes[j] = false;
                }
            }
        }
    }

    @Override
    public void run()
    {
        // do sieves here.
        // each thread will sieve the range of numbers in multiples of
        // its starting point
        SieveEmForGood();
    }
}

File: python/multi-threading/primes.py
from threading import Thread
from multiprocessing import Process, Pool, Value, Array
import multiprocessing as mp
import concurrent.futures
from typing import Final
import numpy as np
import time
import math


# UPPER_BOUND: Final = int(1e8)
NUM_THREADS: Final = 8
# cnt = 0
# lock = None
# primes = np.ones(UPPER_BOUND+1, dtype=np.bool)

class Siever:

    def __init__(self):
        # self.cnt = 0
        pass

    # def init(self, l):
    #     global lock
    #     lock = l

    def sieveSingleThread(self, primes: np.ndarray, startInd: int, UPPER_BOUND: int) -> None:
        while startInd * startInd <= UPPER_BOUND:
            if primes[startInd]:
                primes[startInd*startInd :: startInd] = False
            startInd += 2

    # 8 threads will count all the primes in its segment
    # for 1e8, the first thread will count from 0 to 12.5 million, exclusive;
    # the second thread will count from 12.5 million to 25 million, exclusive;
    # and so on and so forth
    def count_and_add_primes(self, args):
        start, primes, cnt = args[0], args[1], args[2]
        offset = 12500000
        for i in range(start+1, start+offset, 2):
                cnt.value += 1
                print(cnt.value)





if __name__ == '__main__':

    # starting points for the sieve, one for each thread
    # primeNums = np.array([3, 5, 7, 11, 13, 17, 31, 41])

    # primeNums = np.array([3, 5, 7, 11, 13, 17, 31])

    UPPER_BOUND = int(1e8)
    primes = np.ones(UPPER_BOUND+1, dtype=np.bool)
    # 0 and 1 are not prime numbers
    primes[0], primes[1] = False, False
    # atomic integer
    # primeCounter = Value('i', 1)
    atomicCnt = Value('i', 0)

    s = Siever()
    l = mp.Lock()
    start = time.time()

    # for num in primeNums:
    s.sieveSingleThread(primes, 3, UPPER_BOUND)

    # results = [2]
    # total = 2
    # for i in range(3, UPPER_BOUND, 2):
    #     if primes[i]:
    #         results.append(i)
    #         total += i


    lstOfArgs = [
        (0, primes, atomicCnt),
        (12500000, primes, atomicCnt),
        (25000000, primes, atomicCnt),
        (37500000, primes, atomicCnt),
        (50000000, primes, atomicCnt),
        (62500000, primes, atomicCnt),
        (75000000, primes, atomicCnt),
        (87500000, primes, atomicCnt),
        ]

    # shared_array = mp.Array('i', primes)


    pool = Pool(processes=8)
    result = pool.map(s.count_and_add_primes, lstOfArgs)
    pool.close()
    pool.join()

    print('{} ms'.format((time.time() - start) * 1e3))

    # print('total number of primes under 10^8: {}'.format(len(results)))
    print('total sum of all primes under 10^8: {}'.format(s.cnt))

    # print('largest 10 primes: {}'.format(results[-10::]))

    print('done')










    pass


